app/controllers/api/v1/base_controller.rb

```diff
@@ -8,20 +8,34 @@ class BaseController < ApplicationController
       private
 
       def require_authentication!
-        user = User.find(request.headers["X-User-Id"])
+        authenticator = Authenticator.new(
+          user: User.find_by(id: request.headers["X-User-Id"]),
+          provided_token: request.headers["Authorization"]
+        )
 
-        if user.present? && valid_token?(user.authentication_token)
-          sign_in user
+        if authenticator.valid?
+          sign_in authenticator.user
         else
-          render status: :unauthorized
+          head :unauthorized
         end
       end
 
-      def valid_token?(token)
-        ActiveSupport::SecurityUtils.secure_compare(
-          "Token #{token}",
-          request.headers["Authorization"]
-        )
+      class Authenticator
+        attr_reader :user, :provided_token
+
+        def initialize(user:, provided_token:)
+          @user = user
+          @provided_token = provided_token
+        end
+
+        def valid?
+          return false unless user && provided_token
```

Comment:
following on from: https://github.com/abtion/muffi/pull/51#discussion_r415589660

I was a bit surprised that this line ended up in the refactored version too:
`          return false unless user && provided_token`

But here it doesn't bug me as much.

It's still a guard clause, but it's guarding against 2 things that are in the same scope, which is as arguments passed into our constructor. 
They are 2 things which are fundamental to the authentication, that there is a user and that there is a provided token that we need to compare to.

I used `provided_token` here to more clearly differentiate from the `authentication_token` which is a property of the user model.

Reply:
You make a very clear point, I like it 👍